### PR TITLE
feat: Support templateRef in LifecycleHook. Fixes #7558

### LINKED ISSUE #7558: Errors when trying to use templateRef on hooks.exit
## Summary

What happened/what you expected to happen?

When trying to use hooks.exit similar to [this example](https://github.com/argoproj/argo-workflows/blob/master/examples/exit-handler-with-param.yaml), but replacing template with templateRef - getting a runtime error on the exit step with a message: "failed to get a template".

Also, when opening the manifest of the exit step there is a weird exception:

image

The workflow hangs on running state.

The manifest for the top level node shows that the template on the hooks.exit level is empty.

**What version of Argo Workflows are you running?**

V3.2.4

**What Kubernetes provider are you using?**

minikube

**Message from the maintainers**:

Impacted by this bug? Give it a 👍. We prioritise the issues with the most 👍.

### PR DESCRIPTION
**Pull request by dinever**
Signed-off-by: Peixuan Ding <dingpeixuan911@gmail.com>

Hi! This PR adds `templateRef` support in `LifecycleHook`, allowing users to call a workflow template directly from a lifecycle hook. Please see #7558 for details.

Thanks!

### CODE REVIEW

File: pkg/apis/workflow/v1alpha1/workflow_types.go [hunk 1]
```diff
@@ -1199,8 +1199,12 @@ func (lchs LifecycleHooks) GetExitHook() *LifecycleHook {
 }
 
 type LifecycleHook struct {
-	Template  string    `json:"template," protobuf:"bytes,1,opt,name=template"`
+	// Template is the name of the template to execute by the hook
+	Template string `json:"template," protobuf:"bytes,1,opt,name=template"`
+	// Arguments hold arguments to the template
 	Arguments Arguments `json:"arguments,omitempty" protobuf:"bytes,2,opt,name=arguments"`
+	// TemplateRef is the reference to the template resource to execute by the hook
+	TemplateRef *TemplateRef `json:"templateRef,omitempty" protobuf:"bytes,3,opt,name=templateRef"`
```
Comment (dinever) [inline pkg/apis/workflow/v1alpha1/workflow_types.go:1207]: Here's the main change.

**Comment by sarabala1979** [comment]
LGTM, added @jessesuen and @alexec for approving `spec` change

File: pkg/apis/workflow/v1alpha1/workflow_types.go (hunk 1 above)
Comment (sarabala1979) [reply pkg/apis/workflow/v1alpha1/workflow_types.go:1207]: good enhancement @dinever

**Review by alexec — COMMENTED** [review]
@sarabala1979 are you happy with the spec changes please?

**Comment by sarabala1979** [comment]
> @sarabala1979 are you happy with the spec changes please?

yes

**Comment by alexec** [comment]
@sarabala1979 if you are happy, I am happy.

**Review by sarabala1979 — APPROVED** [review]

**Pull request merged**